funnel/views/participant.py

```diff
@@ -144,18 +159,66 @@ def participant_badge(profile, space, participant):
     (Profile, {'name': 'profile'}, 'g.profile'),
     ((ProposalSpace, ProposalSpaceRedirect), {'name': 'space', 'profile': 'profile'}, 'space'),
     (Event, {'name': 'name', 'proposal_space': 'space'}, 'event'),
-    (Participant, {'id': 'participant_id'}, 'participant'),
     permission='checkin-event')
-def event_checkin(profile, space, event, participant):
-    attendee = Attendee.get(event, participant)
+def event_checkin(profile, space, event):
     form = forms.Form()
     if form.validate_on_submit():
-        # Toggle check-in status
-        attendee.checked_in = not attendee.checked_in
+        checked_in = True if request.form.get('checkin') == 't' else False
+        participant_ids = request.form.getlist('pid')
+        for participant_id in participant_ids:
+            participant = Participant.query.filter_by(id=participant_id).first()
+            attendee = Attendee.get(event, participant)
+            attendee.checked_in = checked_in
         db.session.commit()
+        if request.is_xhr:
+            return jsonify(status=True, participant_ids=participant_ids, checked_in=checked_in)
     return redirect(url_for('event', profile=space.profile.name, space=space.name, name=event.name), code=303)
 
 
+@app.route('/<space>/event/<name>/checkin/api', methods=['POST'], subdomain='<profile>')
+@lastuser.requires_login
+@load_models(
+    (Profile, {'name': 'profile'}, 'g.profile'),
+    ((ProposalSpace, ProposalSpaceRedirect), {'name': 'space', 'profile': 'profile'}, 'space'),
+    (Event, {'name': 'name', 'proposal_space': 'space'}, 'event'),
+    permission='checkin-event')
+def checkin_puk(profile, space, event):
+    checked_in = True if request.form.get('checkin') == 't' else False
+    participant_puks = request.form.getlist('puk')
+    attendees = []
+    for participant_puk in participant_puks:
+        participant = Participant.query.filter_by(puk=participant_puk).first()
+        attendee = Attendee.get(event, participant)
```

Comment:
As above, looping is inefficient.

Reply:
@jace Update it to attendee = Attendee.get(event, participant_id), would that be fine?